billy-gin/src/main/java/com/premiumminds/billy/gin/services/impl/pdf/FOPPDFTransformer.java

```diff
@@ -0,0 +1,152 @@
+		String rval = "<" + node.getKey();
+		if (null == node.getValue() && !node.hasChildren()) {
+			rval += "/>";
+		} else {
+			rval += ">";
+			rval += " " + (null != node.getValue() ? StringEscapeUtils.escapeXml(node.getValue()) : "");
+			for (Node<String, String> child : node.getChildren()) {
+				rval += FOPPDFTransformer.toXML(child);
+			}
+			rval += "</" + node.getKey() + ">";
+		}
+		return rval;
+	}
+
+	protected OutputStream transformToStream(InputStream templateStream,
+			ParamsTree<String, String> documentParams,
+			OutputStream outStream)
+		throws ExportServiceException {
+
+		// the XML file from which we take the name
+		Source source = this.mapParamsToSource(documentParams);
+		// creation of transform source
+		StreamSource transformSource = new StreamSource(templateStream);
+
+		// create an instance of fop factory
+		FopFactory fopFactory = FopFactory.newInstance();
+		// a user agent is needed for transformation
+		FOUserAgent foUserAgent = fopFactory.newFOUserAgent();
+		// to store output
+
+		Transformer xslfoTransformer = this.getTransformer(transformSource);
+		try {
+			// Construct fop with desired output format
+			Fop fop = fopFactory.newFop(MimeConstants.MIME_PDF, foUserAgent,
+					outStream);
+			
+			// Resulting SAX events (the generated FO)
+			// must be piped through to FOP
+			Result res = new SAXResult(fop.getDefaultHandler());
+
+			// Start XSLT transformation and FOP processing
+			// everything will happen here..
+			xslfoTransformer.transform(source, res);
+			
+			// if you want to get the PDF bytes, use the following code
+			return outStream;
+		} catch (FOPException e) {
+			throw new ExportServiceException(
+					"Error using FOP to open the template", e);
+		} catch (TransformerException e) {
+			throw new ExportServiceException(
+					"Error generating pdf from template and data source", e);
+		}
+	}
+
+	public File toFile(URI fileURI, InputStream templateStream,
+			ParamsTree<String, String> documentParams) throws ExportServiceException {
+		// if you want to save PDF file use the following code
+		File pdffile = new File(fileURI);
+		OutputStream out;
+		try {
+			out = new FileOutputStream(pdffile);
+			out = new java.io.BufferedOutputStream(out);
+			FileOutputStream str = new FileOutputStream(pdffile);
+			this.transformToStream(templateStream, documentParams, out);
+			str.close();
+			out.close();
+			return pdffile;
+		} catch (FileNotFoundException e) {
+			throw new ExportServiceException("Could not create pdf file", e);
+		} catch (IOException e) {
+			throw new ExportServiceException(
+					"IO error while saving the pdf file", e);
+		}
+	}
+
+	private Transformer getTransformer(StreamSource streamSource) {
+		// setup the xslt transformer
+		net.sf.saxon.TransformerFactoryImpl impl = new net.sf.saxon.TransformerFactoryImpl();
+
+		try {
+			return impl.newTransformer(streamSource);
+		} catch (TransformerConfigurationException e) {
+			e.printStackTrace();
+		}
+		return null;
```

Comment:
`return null`  :thinking:  new factory instance in each call? Could it be a static final?